lib/backend/k8s/resources/resources.go

```diff
@@ -100,3 +102,74 @@ func SetCalicoMetadataFromK8sAnnotations(calicoRes Resource, k8sRes Resource) {
 		}
 	}
 }
+
+// Store Calico Metadata in the in the k8s resource annotations for CRD backed resources.
+// This should store all Metadata except for those stored in Annotations and Labels and
+// store them in annotations.
+func StoreCalicoMetadataInK8sAnnotations(res Resource) error {
+	metadataBytes, err := json.Marshal(res.GetObjectMeta().(*metav1.ObjectMeta))
+	if err != nil {
+		return err
+	}
+	annotations := res.GetObjectMeta().GetAnnotations()
+	if len(annotations) == 0 {
+		annotations = make(map[string]string)
+	}
+	annotations[metadataAnnotation] = string(metadataBytes)
+	res.GetObjectMeta().SetAnnotations(annotations)
+
+	// Make sure to clear out all of the Calico Metadata out of the ObjectMeta except for
+	// Name, Namespace, and ResourceVersion.
+	res.GetObjectMeta().SetSelfLink("")
+	res.GetObjectMeta().SetUID("")
+	res.GetObjectMeta().SetGeneration(int64(0))
+	var emptyTime metav1.Time
+	res.GetObjectMeta().SetCreationTimestamp(emptyTime)
+	res.GetObjectMeta().SetDeletionTimestamp(nil)
+	res.GetObjectMeta().SetDeletionGracePeriodSeconds(nil)
+	res.GetObjectMeta().SetInitializers(nil)
+	var emptyStringSlice []string
+	res.GetObjectMeta().SetFinalizers(emptyStringSlice)
+	var emptyReferences []metav1.OwnerReference
+	res.GetObjectMeta().SetOwnerReferences(emptyReferences)
+	res.GetObjectMeta().SetClusterName("")
+
+	return nil
+}
+
+// Retrieve all of the Calico Metadata from the k8s resource annotations for CRD backed
+// resources. This should remove the relevant Calico Metadata annotation when it has finished.
+func GetCalicoMetadataFromK8sAnnotations(res Resource) error {
+	defer CleanAnnotations(res)
+	annotations := res.GetObjectMeta().GetAnnotations()
+	if len(annotations) == 0 {
+		return errors.New("Attempting to create a resource from a resource without Calico metadata set in the annotations.")
+	}
+	meta := &metav1.ObjectMeta{}
+	err := json.Unmarshal([]byte(annotations[metadataAnnotation]), meta)
+	if err != nil {
+		return err
+	}
+
+	// Overwrite the K8s metadata with the Calico metadata.
+	res.GetObjectMeta().SetSelfLink(meta.SelfLink)
+	res.GetObjectMeta().SetUID(meta.UID)
+	res.GetObjectMeta().SetGeneration(meta.Generation)
+	res.GetObjectMeta().SetCreationTimestamp(meta.CreationTimestamp)
+	res.GetObjectMeta().SetDeletionTimestamp(meta.DeletionTimestamp)
+	res.GetObjectMeta().SetDeletionGracePeriodSeconds(meta.DeletionGracePeriodSeconds)
+	res.GetObjectMeta().SetInitializers(meta.Initializers)
+	res.GetObjectMeta().SetFinalizers(meta.Finalizers)
+	res.GetObjectMeta().SetOwnerReferences(meta.OwnerReferences)
+	res.GetObjectMeta().SetClusterName(meta.ClusterName)
+
+	return nil
+}
+
+// Cleans out the Calico metadata from the annotations field. Needed to reduced side effects
+// that come from needing to modify the input data.
+func CleanAnnotations(res Resource) {
```

Comment:
This doesn't need to be public.

It could also benefit from a better name - I wasn't sure what it was really doing at first.

Reply:
Maybe something like :   removeCalicoMetadataAnnotation

Reply:
IN fact I think I would just scrap this method and build it into the method above - that way those two methods are more symmetric.

One updates Metadata for K8s
One updates Metadata for Calico

Reply:
With that in mind, perhaps naming those methods something like:
```
ConvertMetadataCalicoToK8s()
ConvertMetadataK8sToCalico()
```

Reply:
Actually, do we need these to be two separate methods?

Can't we just include this processing in the ConvertResourceToKVPair and ConvertKVPairToResource functions.

Reply:
Hmm, given that this will only be used by the CRD backed resources, I wonder if this should just be part of the CRD conversion methods.  Don't mind it staying here though if you prefer but you should tighten up the method names though to indicate what they're doing.

Reply:
This logic used to be built into the above methods but then I realized that it also needed to be present in the `customK8sResourceClient` if we wanted to ensure that the `resource` input is not modified (since ideally functions should not leave side effects on the inputs). If we're fine with the input data being modified, then I can move the logic back into the above methods and fix the tests. I was really only made aware of this from the new k8s tests.